Library/Homebrew/livecheck/livecheck.rb

```diff
@@ -0,0 +1,431 @@
+
+        current = formula.head? ? formula.installed_version.version.commit : formula.version
+
+        latest = if formula.stable?
+          version_info = latest_version(formula, args: args)
+          version_info[:latest] if version_info.present?
+        else
+          formula.head.downloader.fetch_last_commit
+        end
+
+        if latest.blank?
+          no_versions_msg = "Unable to get versions"
+          raise TypeError, no_versions_msg unless args.json?
+
+          next version_info if version_info.is_a?(Hash) && version_info[:status] && version_info[:messages]
+
+          next status_hash(formula, "error", [no_versions_msg], args: args)
+        end
+
+        if (m = latest.to_s.match(/(.*)-release$/)) && !current.to_s.match(/.*-release$/)
+          latest = Version.new(m[1])
+        end
+
+        # A HEAD-only formula is outdated when the latest commit hash and installed commit hash differ
+        is_outdated = if formula.head?
+          (current != latest)
+        else
+          (current < latest)
+        end
+
+        is_newer_than_upstream = formula.stable? && (current > latest)
+
+        info = {
+          formula: formula_name(formula, args: args),
+          version: {
+            current:             current.to_s,
+            latest:              latest.to_s,
+            outdated:            is_outdated,
+            newer_than_upstream: is_newer_than_upstream,
+          },
+          meta:    {
+            livecheckable: formula.livecheckable?,
+          },
+        }
+        info[:meta][:head_only] = true if formula.head?
+        info[:meta].merge!(version_info[:meta]) if version_info.present? && version_info.key?(:meta)
+
+        next if args.newer_only? && !info[:version][:outdated]
+
+        has_a_newer_upstream_version ||= true
+
+        if args.json?
+          info.except!(:meta) unless args.verbose?
+          next info
+        end
+
+        print_latest_version(info, args: args)
+        nil
+      rescue => e
+        Homebrew.failed = true
+
+        if args.json?
+          status_hash(formula, "error", [e.to_s], args: args)
+        elsif !args.quiet?
+          onoe "#{Tty.blue}#{formula_name(formula, args: args)}#{Tty.reset}: #{e}"
+          nil
+        end
+      end
+
+      if [args.newer_only?, !has_a_newer_upstream_version, !args.json?, !args.debug?].all?
+        puts "No newer upstream versions."
+      end
+
+      puts JSON.generate(formulae_checked.compact) if args.json?
+    end
+
+    # Returns the fully-qualified name of a formula if the full_name argument is
+    # provided, returns the name otherwise.
+    # @return [String]
+    def formula_name(formula, args:)
+      args.full_name? ? formula.full_name : formula.name
+    end
+
+    def status_hash(formula, status_str, messages = nil, args:)
+      status_hash = {
+        formula: formula_name(formula, args: args),
+        status:  status_str,
+      }
+      status_hash[:messages] = messages.presence
```

Comment:
Would
```ruby
status_hash[:messages] = messages.present? ? messages : []
```
be a better option? However, I'm not sure whether the intention was to have the messages key in `status_hash` no matter what.

Reply:
@MikeMcQuaid, what are your thoughts here?

I still prefer the original `status_hash[:messages] = messages if messages.is_a?(Array)` code, as it only sets `status_hash[:messages]` when it's appropriate. `messages.present?` and `messages.presence` are less strict, since they're only checking for a non-blank value (not ensuring `messages` is an array).

If Mike's intention is to always have `status_hash[:messages]`, then I would probably prefer your initial version of this, as there's less work involved (it's ~30% faster than `messages.presence.to_a`): `status_hash[:messages] = messages.present? ? messages : []`.

Reply:
I don't mind either way. I don't think performance matters here as it's not a tight loop.

Reply:
Reverting this to the original `is_a?(Array)` code.